### ISSUE TITLE
`ClassVar[Callable]` referenced from `self` - false positive

### ISSUE THREAD

**Issue by beauxq** [microsoft/pyright#8242]
This has correct type annotations and runs without error, but pyright reports the wrong number of arguments for the `a_maker` function.

```python
from typing import Callable, ClassVar


class A:
    def __init__(self, x: int, y: str) -> None:
        print(x, y)


class B:
    a_maker: ClassVar[Callable[[int, str], A]] = A

    def f(self) -> None:
        self.a_maker(3, "3")  # Expected 1 positional argument  (reportCallIssue)


B().f()
```

https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCAhgDYlEBGJApgDSHkDOjAakSAFBcDGTjUAQQBcHKGKgATasCgB9WaiQx5ACkbUSwegA8hmFDHpw9jGCACUUALQA%2BKADkwKaiPFuoCEKhgrtR81wcvETMUABCruJEshBEANbUIHoEfGwgANrEZJQ06ene9KYgALr0AsXFUAC8glxuUjLAahrAlrYOTi6i7mLqmgB00bEJICoAzPQARGOTAVxhKub9TQFAA

pyright version 1.1.368

**Comment by erictraut** [comment]
Pyright's behavior is correct here, so this isn't a bug.

You've defined a class variable that contains a `Callable`. There's currently no way in the Python type system to distinguish between a callable that is a staticmethod, classmethod or instance method. All three of these are class variables, but they have different behaviors. Pyright (and other type checkers like mypy) assume that if you assign a value to a Callable class variable, it should be interpreted as an instance method — one that requires binding to `self`.

The Python typing spec is currently silent on how type checkers should work in this case, but all of the major type checkers are in alignment in their behavior. Pyright and mypy generate the same error in this case. 

We are actively working to fill in gaps like this in the Python typing spec. If and when such behavior is specified, pyright will honor the spec. Until then, I don't plan to change this behavior.

You can work around this issue by doing the following:

```python
    def f(self) -> None:
        B.a_maker(3, "3")
```

or

```python
    def f(self) -> None:
        type(self).a_maker(3, "3")
```
